File: ch6/akkademy-db-scala/src/main/scala/com/akkademy/AkkademyDb.scala
package com.akkademy

import akka.actor._
import akka.cluster.Cluster
import akka.cluster.ClusterEvent.{MemberEvent, UnreachableMember}
import akka.event.Logging
import com.akkademy.messages.{GetRequest, KeyNotFoundException, SetRequest}

import scala.collection.mutable.HashMap

class AkkademyDb extends Actor {
  val map = new HashMap[String, Object]
  val log = Logging(context.system, this)

  override def receive = {
    case SetRequest(key, value) =>
      log.info("received SetRequest - key: {} value: {}", key, value)
      map.put(key, value)
      sender() ! Status.Success
    case GetRequest(key) =>
      log.info("received GetRequest - key: {}", key)
      val response: Option[Object] = map.get(key)
      response match{
        case Some(x) => sender() ! x
        case None => sender() ! Status.Failure(new KeyNotFoundException(key))
      }
    case o => Status.Failure(new ClassNotFoundException)
  }
}


class ClusterController extends Actor {
  val log = Logging(context.system, this)
  val cluster = Cluster(context.system)

  var members = List(ActorRef)

  override def preStart() {
    cluster.subscribe(self, classOf[MemberEvent], classOf[UnreachableMember])
  }

  override def postStop() {
    cluster.unsubscribe(self)
  }

  override def receive = {
    case x: MemberEvent => log.info("MemberEvent: {}", x)
    case x: UnreachableMember => log.info("UnreachableMember {}: ", x)
  }
}


object Main extends App {
  val system = ActorSystem("Akkademy")
  val clusterController = system.actorOf(Props[ClusterController], "clusterController")

}

File: ch6/akkademy-db-java/src/main/java/com/akkademy/AkkademyDb.java
package com.akkademy;

import akka.actor.AbstractActor;
import akka.actor.Status;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import com.akkademy.japi.GetRequest;
import com.akkademy.japi.KeyNotFoundException;
import com.akkademy.japi.SetRequest;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import java.util.HashMap;
import java.util.Map;

public class AkkademyDb extends AbstractActor {
    protected final LoggingAdapter log = Logging.getLogger(context().system(), this);
    protected final Map<String, Object> map = new HashMap<String, Object>();

    private AkkademyDb(){
        receive(ReceiveBuilder.
                        match(SetRequest.class, message -> {
                            log.info("Received Set request: {}", message);
                            map.put(message.key, message.value);
                            sender().tell(new Status.Success(message.key), self());
                        }).
                        match(GetRequest.class, message -> {
//                            log.info("Received Get request: {}", message);
                            Object value = map.get(message.key).toString() +
                            map.get(message.key) +
                            map.get(message.key) +
                            map.get(message.key) +
                            map.get(message.key) +
                            map.get(message.key).toString() + "HELLO";
                            value.toString().contains("HELO");
                            value.toString().contains("HELO");
                            value.toString().contains("HLLO");
                            value.toString().contains("HELLO");
                            value.toString().contains("HELO");
                            value.toString().contains("HELO");
                            value.toString().contains("HELO");
                            value.toString().contains("HLLO");
                            value.toString().contains("HELL");
                            value.toString().endsWith("HELL");
                            value.toString().endsWith("HELL");
                            value.toString().endsWith("HLL");
                            value.toString().endsWith("ELL");
                            value.toString().endsWith("HEL");
                            value.toString().endsWith("HE");
                            value.toString().endsWith("HE");
                            value.toString().endsWith("HEzzz");
                            value.toString().endsWith("HEzzzz");

                            value.hashCode();
                            (value + "hello").hashCode();
                            (value + "he5l8lo").hashCode();
                            (value + "hel5lo").hashCode();
                            (value + "h5ello").hashCode();
                            (value + "hel7lo").hashCode();

                            Object response = (value != null)
                                    ? value
                                    : new Status.Failure(new KeyNotFoundException(message.key));
                            sender().tell(response, self());
                        }).
                        matchAny(o ->
                                sender().tell(new Status.Failure(new ClassNotFoundException()), self())
                        ).build()
        );
    }
}